### PR TITLE
Native interpreter integration

### PR DESCRIPTION
**Pull request by foolnotion**
This updates the native interpreter with the following features

- removed dependency on Ceres lib (use Eigen::LevenbergMarquardt and Ceres jets separately)
- support weighting of residuals
- improvements in code quality

### CODE REVIEW

**Review by gkronber — APPROVED** [review]
The new results look much better:
First the results for the large function set: (+, *,/, sin, cos, tanh, log, exp, abs, aq, sqr, sqrt, cube, cbrt). Results for the reduced set are below.

On the training set we get better or equal results with the native param opt (black) compared to managed (red). The improved training quality is a bit surprising but a nice benefit. The reason for the improvement is not clear. It is likely to be the minpack optimizer compared to the alglib optimizer. There is also the difference that native paramopt optimizes for MSE directly and does not use linear scaling. There could be a potential improvement for the native optimizer by enforcing linear scaling through the grammar. We should make an experiment on this in the future.
training_full


On the test set I don't see a difference. However, for these settings (50nodes) the results tend to overfit causing large variance. Tuning parameters for each problem individually (and shuffling the data for each repetition) may even show better test results for the native paramOpt. This is left for a future experiment.
 test_full

Execution time is much better (speeup > 5 for chemical, tower, housing, poly10, ...).
This is incredibly impressive considering that large parts of the GP pipeline are still unoptimized. However, the results are for a single-thread, with multiple threads the improvement will not be as large as the sequential part remains. I'm very happy to finally have this directly in HL.

execution_time_full

speedup_full


Simple function set (+,-,*,/): Similar training results as above. There is a single problem 522_pm10 where the managed results are better.
training_reduced

No difference on test set
test_reduced

Execution time:
executiong_time_reduced

For the reduced function set the speedup is even larger
speedup_reduced


In summary: impressive work. This is a bit improvement for the HL SymReg implementation. Thank you Bogdan.

Future work:
 - Compare test results using the PennML testing protocol.
 - Test if native paramOpt results can be improved by using a linear scaling grammar
 - Completely switch to native paramOpt and native interpreter as a default. Remove all old code (AutoDiff, managed ParamOpt, almost all of the interpreters
